### PR TITLE
[REF] Volumes management

### PR DESCRIPTION
**Pull request by YannickB**
Hello all,

This is the PR for #219. As announced, this is a heavy refactor of the container structure, to remove the data/files/exec we used until now.

-Add new clouder.volume model. The clouder.volume is linked to service but not removed when service is unlink (unless explicit)
-Bind/Postfix/Proxy/Postgres/Odoo templates are simplified and don't use data/files/exec anymore
-SSH template is now a new application type
-Resolve some bug with raise_error having wrong number of argument
-Service now know they have to create a container thanks to an explicit checkbox, remove dummy field.

Also I officially drop the support for Odoo 8. You can still deploy Odoo 8 instance with Clouder, but you can't install Clouder module on Odoo 8 anymore. The reason is mainly the use of odoo tag in xml file.

Some image use my yannickburon docker hub repo, until the PR is merged and I can create the new image on official repo.

### CODE REVIEW

File: clouder_template_odoo/images/odoo/Dockerfile [hunk 1]
```diff
@@ -41,9 +41,17 @@ RUN apk add --update --no-cache postgresql-client nodejs py2-pip \
...
+        # && apk add --update --no-cache --repository http://dl-cdn.alpinelinux.org/alpine/edge/testing wkhtmltopdf \
         && apk del .build-deps
 
+RUN apk add --update --no-cache \
+    libgcc libstdc++ libx11 glib libxrender libxext libintl \
+    libcrypto1.0 libssl1.0 \
+    ttf-dejavu ttf-droid ttf-freefont ttf-liberation ttf-ubuntu-font-family
+
+RUN wget https://github.com/madnight/docker-alpine-wkhtmltopdf/raw/master/wkhtmltopdf -O /usr/bin/wkhtmltopdf
+RUN chmod +x /usr/bin/wkhtmltopdf
```
Comment (lasley) [inline clouder_template_odoo/images/odoo/Dockerfile]: Take a look at this - https://github.com/LasLabs/docker-odoo/blob/master/Dockerfile#L45

Comment (YannickB) [inline clouder_template_odoo/images/odoo/Dockerfile]: @lasley For the record, your Dockerfile use a wkhtmlbinary build without qt support, hence no header/footer and other random trouble. That's why I tried to find a binary on Internet compiled with both qt and alpine, this works on my instance.

Hope wkhtmltopdf provide soon a package for alpine with qt, this is quite bothersome.

**Comment by lasley** [comment]
@tedsalmon please review

File: clouder/clouder_template_ssh/template.xml [hunk 2]
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
```
Comment (lasley) [inline clouder/clouder_template_ssh/template.xml]: No need for data tag using new odoo tag, unless you need both update and noupdate data in the same file

File: clouder/images/php/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+RUN echo "" >> /etc/supervisord.conf
+RUN echo "[supervisord]" >> /etc/supervisord.conf
+RUN echo "nodaemon=true" >> /etc/supervisord.conf
+RUN echo "" >> /etc/supervisord.conf
+RUN echo "[program:nginx]" >> /etc/supervisord.conf
+RUN echo "command=nginx" >> /etc/supervisord.conf
+RUN echo "[program:php]" >> /etc/supervisord.conf
+RUN echo "command=php-fpm7" >> /etc/supervisord.conf
+
+CMD supervisord -c /etc/supervisord.conf
```
Comment (lasley) [inline clouder/images/php/Dockerfile]: newline at EOF

File: clouder/images/php/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    php7-intl php7-bcmath php7-mbstring php7-dom php7-xmlreader mysql-client openssh-client
+
+RUN echo "" >> /etc/supervisord.conf
+RUN echo "[supervisord]" >> /etc/supervisord.conf
+RUN echo "nodaemon=true" >> /etc/supervisord.conf
+RUN echo "" >> /etc/supervisord.conf
+RUN echo "[program:nginx]" >> /etc/supervisord.conf
+RUN echo "command=nginx" >> /etc/supervisord.conf
+RUN echo "[program:php]" >> /etc/supervisord.conf
+RUN echo "command=php-fpm7" >> /etc/supervisord.conf
```
Comment (lasley) [inline clouder/images/php/Dockerfile]: IMO these would be more clear in a file, then use `cat $FILE > /etc/supervisord.conf`

File: clouder/images/ssh/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+MAINTAINER Yannick Buron yannick.buron@gmail.com
+
+RUN touch /tmp/odoo-ssh
+RUN apk add --update openssh
+RUN mkdir /var/run/sshd
+RUN chmod 0755 /var/run/sshd
+RUN /usr/bin/ssh-keygen -A
+USER root
+
+CMD /usr/sbin/sshd -ddd
```
Comment (lasley) [inline clouder/images/ssh/Dockerfile]: newline EOF

File: clouder/models/service.py [hunk 6]
```diff
@@ -1309,45 +1320,71 @@ def deploy_post(self):
...
+    def recursive_deploy_volume(self):
+        """
+        Recursively deploy services one by one
+        """
+
+        # Create Volume
+        if self.container:
+            volume_obj = self.env['clouder.volume']
+            for service_volume in self.volume_ids:
+                volume_name = self.name + '-' + service_volume.name
```
Comment (lasley) [inline clouder/models/service.py]: `'%s-%s' % (self.name, service_volume.name)`

File: clouder/models/service.py [hunk 7]
```diff
@@ -1309,45 +1320,71 @@ def deploy_post(self):
...
+                    volumes = volume_obj.search([('name', '=', volume_name)])
+                    if not volumes:
+                        service_volume.volume_id = volume_obj.create(
+                            {'name': volume_name,
+                             'path': service_volume.localpath,
+                             'node_id': self.node_id.id})
+                    else:
+                        service_volume.volume_id = volumes[0]
+
+        for child in self.child_ids:
```
Comment (lasley) [inline clouder/models/service.py]: `self.child_ids.filtered(lambda r: r.child_id)` & can remove below conditional

File: clouder/models/service.py [hunk 8]
```diff
@@ -1309,45 +1320,71 @@ def deploy_post(self):
...
     def recursive_deploy_one(self):
         """
         Recursively deploy services one by one
         """
+        if self.container:
+            self.hook_deploy_one()
 
-        if self.child_ids:
-            for child in self.child_ids:
+        for child in self.child_ids:
```
Comment (lasley) [inline clouder/models/service.py]: use filtered & remove below conditional

File: clouder/models/service.py [hunk 9]
```diff
@@ -1309,45 +1320,71 @@ def deploy_post(self):
...
-            for child in self.child_ids:
-                child.child_id.recursive_deploy_post()
-        else:
-            # We only call the hook if service without children
+        if self.container:
             self.deploy_post()
-            self.deploy_links()
+        self.deploy_links()
+
+        for child in self.child_ids:
```
Comment (lasley) [inline clouder/models/service.py]: `.filtered(`

File: clouder/models/service.py [hunk 10]
```diff
@@ -1309,45 +1320,71 @@ def deploy_post(self):
...
-                child.child_id.recursive_backup()
-        else:
-            # Avoid backup alert on monitoring
-            self = self.with_context(backup_comment='First backup')
-            self.backup_exec(no_enqueue=True)
+        # Avoid backup alert on monitoring
+        self = self.with_context(backup_comment='First backup')
+        self.backup_exec(no_enqueue=True)
+
+        for child in self.child_ids:
```
Comment (lasley) [inline clouder/models/service.py]: `self.child_ids.mapped('child_id').recursive_backup()`

File: clouder/models/service.py [hunk 11]
```diff
@@ -1362,13 +1399,10 @@ def deploy(self):
         # Create childs
         if self.child_ids:
             for child in self.child_ids:
```
Comment (lasley) [inline clouder/models/service.py]: Use filtered

File: clouder/models/service.py [hunk 12]
```diff
@@ -1415,6 +1452,11 @@ def hook_purge_one(self):
         return
 
     @api.multi
+    def purge_with_volume(self):
+        self = self.with_context(no_enqueue=True, purge_volume=True)
```
Comment (lasley) [inline clouder/models/service.py]: Reassigning self is pretty hacky IMO. Why not just assign to another variable and use that?

File: clouder/models/service.py [hunk 13]
```diff
@@ -1423,19 +1465,24 @@ def purge(self):
...
-        else:
-            # If not compose, purge current service
-            if not self.compose:
+            if self.container:
                 self.stop()
                 # self.purge_salt()
                 self.hook_purge_one()
+
+        if 'purge_volume' in self.env.context:
+            for volume in self.volume_ids:
```
Comment (lasley) [inline clouder/models/service.py]: `self.volume_ids.mapped('volume_id').unlink()`

File: clouder/images/php/Dockerfile (hunk 4 above)
Comment (YannickB) [inline clouder/images/php/Dockerfile]: I agree but I'm not sure I can properly test it right now. If you don't mind, I'd rather merge it like that and correct all the Dockerfile later.

Comment (lasley) [inline clouder/images/php/Dockerfile]: I'm good with it 👍

**Comment by YannickB** [comment]
Thanks @lasley ! Ready for review

**Pull request merged**